sentry-core/src/metrics.rs

```diff
@@ -0,0 +1,404 @@
+            let mut values = components.next()?.split(':');
+            let name = values.next()?;
+
+            let ty: MetricType = components.next().and_then(|s| s.parse().ok())?;
+            let mut value = mk_value(ty, values.next()?)?;
+
+            for value_s in values {
+                value.merge(mk_value(ty, value_s)?).ok()?;
+            }
+
+            let mut tags = "";
+            let mut timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap_or_default()
+                .as_secs();
+
+            for component in components {
+                if let Some(component_tags) = component.strip_prefix('#') {
+                    tags = component_tags;
+                } else if let Some(component_timestamp) = component.strip_prefix('T') {
+                    timestamp = component_timestamp.parse().ok()?;
+                }
+            }
+
+            Some((
+                BucketKey {
+                    timestamp,
+                    ty,
+                    name: name.into(),
+                    tags: tags.into(),
+                },
+                value,
+            ))
+        }
+
+        if let Some(parsed_metric) = parse(metric) {
+            let _ = self.sender.send(Task::SendMetrics(parsed_metric));
+        }
+    }
+
+    pub fn flush(&self) {
+        let _ = self.sender.send(Task::Flush);
+    }
+
+    fn worker_thread(receiver: Receiver<Task>, shutdown: Arc<AtomicBool>, transport: TransportArc) {
+        let mut buckets = AggregateMetrics::new();
+        let mut last_flush = Instant::now();
+
+        loop {
+            if shutdown.load(Ordering::SeqCst) {
+                Self::flush_buckets(buckets, &transport);
+                return;
+            }
+
+            let timeout = FLUSH_INTERVAL
+                .checked_sub(last_flush.elapsed())
+                .unwrap_or_else(|| Duration::from_secs(0));
+
+            match receiver.recv_timeout(timeout) {
+                Err(RecvTimeoutError::Timeout) | Ok(Task::Flush) => {
+                    // flush
+                    Self::flush_buckets(std::mem::take(&mut buckets), &transport);
+                    last_flush = Instant::now();
+                }
+                Ok(Task::SendMetrics((mut key, value))) => {
+                    // aggregate
+                    let rounding_interval = FLUSH_INTERVAL.as_secs();
+                    let rounded_timestamp = (key.timestamp / rounding_interval) * rounding_interval;
+
+                    key.timestamp = rounded_timestamp;
+
+                    match buckets.entry(key) {
+                        Entry::Occupied(mut entry) => {
+                            let _ = entry.get_mut().merge(value);
+                        }
+                        Entry::Vacant(entry) => {
+                            entry.insert(value);
+                        }
+                    }
+                }
+                _ => {
```

Comment:
```suggestion
                Ok(Task::Shutdown) => {
```

Reply:
Not quite, there is also `RecvTimeoutError::Disconnected`, in which case shutting down is the only viable thing to do.

Reply:
Ah, I didn't catch that `RecvTimeoutError` has two variants.